core/bin/external_node/src/main.rs

```diff
@@ -359,44 +386,204 @@ async fn init_tasks(
             .await
             .context("Failed initializing WS JSON-RPC server")?;
+        app_health.insert_component(ws_server_handles.health_check);
+        task_futures.extend(ws_server_handles.tasks);
+    }
+
+    task_futures.extend(cache_update_handle);
+    task_futures.push(proxy_cache_updater_handle);
+
+    Ok(())
+}
+
+async fn init_tasks(
+    config: &ExternalNodeConfig,
+    connection_pool: ConnectionPool<Core>,
+    main_node_client: HttpClient,
+    task_handles: &mut Vec<task::JoinHandle<anyhow::Result<()>>>,
+    app_health: &AppHealthCheck,
+    stop_receiver: watch::Receiver<bool>,
+    components: &HashSet<Component>,
+) -> anyhow::Result<()> {
+    let mut components = components.clone();
+    let release_manifest: serde_json::Value = serde_json::from_str(RELEASE_MANIFEST)
+        .context("releuse manifest is a valid json document")?;
+    let release_manifest_version = release_manifest["core"].as_str().context(
+        "a release-please manifest with \"core\" version field was specified at build time",
+    )?;
+
+    let version = semver::Version::parse(release_manifest_version)
+        .context("version in manifest is a correct semver format")?;
+    let pool = connection_pool.clone();
+    task_handles.push(tokio::spawn(async move {
+        loop {
+            let protocol_version = pool
+                .connection()
+                .await
+                .unwrap()
+                .protocol_versions_dal()
+                .last_used_version_id()
+                .await
+                .map(|version| version as u16);
+
+            EN_METRICS.version[&(format!("{}", version), protocol_version)].set(1);
+
+            tokio::time::sleep(Duration::from_secs(10)).await;
+        }
+    }));
+
+    let singleton_pool_builder = ConnectionPool::singleton(&config.postgres.database_url);
+
+    // Run the components.
+    let tree_pool = singleton_pool_builder
+        .build()
+        .await
+        .context("failed to build a tree_pool")?;
+
+    if !components.contains(&Component::Tree) {
+        anyhow::ensure!(
+            !components.contains(&Component::TreeApi),
+            "Merkle tree API cannot be started without a tree component"
+        );
+    }
+    // Create a tree reader. If the list of requested components has the tree itself, then
+    // we can get this tree's reader and use it right away. Otherwise, if configuration has
+    // specified address of another instance hosting tree API, create a tree reader to that
+    // remote API. A tree reader is necessary for `zks_getProof` method to work.
+    let tree_reader: Option<Arc<dyn TreeApiClient>> = if components.contains(&Component::Tree) {
+        let tree_api_config = if components.contains(&Component::TreeApi) {
+            Some(MerkleTreeApiConfig {
+                port: config
+                    .tree_component_config
+                    .api_port
+                    .ok_or(anyhow!("should contain tree api port"))?,
```

Comment:
Nit: You can use `.context(_)` on an option with the same effect; it wouldn't require eagerly constructing an error (which could be avoided using `ok_or_else`, but `context` is shorter and arguably more idiomatic).